### PR TITLE
tests: add cross signing tests [WPB-7264]

### PR DESCRIPTION
**Pull request by augustocdias**
# What's new in this PR

This adds 2 new tests for cross signing certificates. 

----
##### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: crypto/src/test_utils/mod.rs [hunk 1]
```diff
@@ -168,31 +174,153 @@ pub async fn run_test_with_deterministic_client_ids<const N: usize>(
...
+pub async fn run_cross_signed_tests_with_client_ids<const N: usize, const F: usize>(
     case: TestCase,
     client_ids: [[&'static str; 3]; N],
+    federated_client_ids: [[&'static str; 3]; F],
+    (domain1, domain2): (&'static str, &'static str),
+    test: impl FnOnce([MlsCentral; N], [MlsCentral; F]) -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + 'static>>
+        + 'static,
+) {
+    assert!(case.is_x509(), "This is only supported for x509 test cases");
+    let _ = pretty_env_logger::try_init();
```
Comment (istankovic) [inline crypto/src/test_utils/mod.rs:203]: Why not just use `pretty_env_logger::init()`, if we're discarding the return value?

File: crypto/src/test_utils/mod.rs [hunk 2]
```diff
@@ -168,31 +174,153 @@ pub async fn run_test_with_deterministic_client_ids<const N: usize>(
...
+    (params1, params2): (CertificateParams, CertificateParams),
+    revoked_display_names: &'static [&'static str],
+) -> (X509TestChain, X509TestChain) {
+    let mut chain1 = init_x509_test_chain(case, &client_ids, revoked_display_names, params1);
+    let mut chain2 = init_x509_test_chain(case, &federated_client_ids, revoked_display_names, params2);
+    chain1.cross_sign(&mut chain2);
+    (chain1, chain2)
+}
+
+pub async fn run_cross_signed_tests_with_client_ids<const N: usize, const F: usize>(
```
Comment (istankovic) [inline crypto/src/test_utils/mod.rs:194]: We should be consistent with other `run_test_` functions, e.g.
```
run_test_with_client_ids_cross_signed_certs
```
(note also: we're running just 1 test, not multiple tests)

File: crypto/src/test_utils/mod.rs [hunk 3]
```diff
@@ -168,31 +174,153 @@ pub async fn run_test_with_deterministic_client_ids<const N: usize>(
...
+pub fn init_federated_x509_test_chains<const N: usize, const F: usize>(
+    case: &TestCase,
+    client_ids: [[&'static str; 3]; N],
+    federated_client_ids: [[&'static str; 3]; F],
+    (params1, params2): (CertificateParams, CertificateParams),
+    revoked_display_names: &'static [&'static str],
+) -> (X509TestChain, X509TestChain) {
+    let mut chain1 = init_x509_test_chain(case, &client_ids, revoked_display_names, params1);
+    let mut chain2 = init_x509_test_chain(case, &federated_client_ids, revoked_display_names, params2);
+    chain1.cross_sign(&mut chain2);
```
[the remark this answers is not in the document]
Comment (augustocdias) [reply crypto/src/test_utils/mod.rs:190]: It is. If you check the `cross_sign` function it adds the intermediates in both of them

File: crypto/src/test_utils/mod.rs (hunk 1 above)
Comment (augustocdias) [reply crypto/src/test_utils/mod.rs:203]: because it can panic if another one is set before

Comment (istankovic) [reply crypto/src/test_utils/mod.rs:203]: Hm, can we ensure that `pretty_env_logger` is initialized only once, say, in some toplevel test fixture or something like that? I see it only being used in two places.

File: crypto/src/test_utils/x509.rs [hunk 4]
```diff
@@ -79,15 +80,50 @@ pub struct X509TestChain {
...
+pub struct X509TestChainArgs {
     pub root_params: CertificateParams,
     pub local_ca_params: CertificateParams,
     pub signature_scheme: SignatureScheme,
-    pub federated_test_chains: &'a [X509TestChain],
     pub local_actors: Vec<X509TestChainActorArg>,
     pub dump_pem_certs: bool,
 }
 
+// Helps debugging certificate chaings by printing the PEM into the stdout
```
Comment (istankovic) [inline crypto/src/test_utils/x509.rs:91]: Typo: `chaings` -> `chains`

File: crypto/src/test_utils/x509.rs [hunk 5]
```diff
@@ -79,15 +80,50 @@ pub struct X509TestChain {
...
+                "{}",
+                &actor
+                    .certificate
+                    .certificate
+                    .to_pem(x509_cert::der::pem::LineEnding::LF)
+                    .unwrap()
+            )
+        })
+    }
+}
```
Comment (istankovic) [inline crypto/src/test_utils/x509.rs:125]: Very useful to have this! :+1:

File: crypto/src/test_utils/x509.rs [hunk 6]
```diff
@@ -253,41 +287,43 @@ impl X509TestChain {
...
+        other_chain.crls = self.crls.clone();
 
-                intermediates.push(cross_signed_intermediate);
-            }
-            let mut federated_actors = federated_chain.actors.clone();
-            for actor in &mut federated_actors {
-                actor.certificate.is_federated = true;
-            }
+        let mut self_new_intermediates = vec![];
+        for fed_intermediate in &other_chain.intermediates {
```
Comment (istankovic) [inline crypto/src/test_utils/x509.rs:305]: `fed_intermediate` can be just `intermediate`.

File: crypto/src/test_utils/x509.rs [hunk 7]
```diff
@@ -253,41 +287,43 @@ impl X509TestChain {
...
+            let cross_signed_intermediate = self.trust_anchor.cross_sign_intermediate(fed_intermediate);
+            self_new_intermediates.push(cross_signed_intermediate);
+        }
 
-            actors.extend(federated_actors);
+        let mut other_new_intermediates = vec![];
+        for fed_intermediate in &self.intermediates {
```
Comment (istankovic) [inline crypto/src/test_utils/x509.rs:311]: Similarly here.

File: crypto/src/test_utils/mod.rs (hunk 1 above)
Comment (istankovic) [reply crypto/src/test_utils/mod.rs:203]: Ah, good point. :+1:

File: crypto/src/e2e_identity/identity.rs [hunk 8]
```diff
@@ -259,6 +259,114 @@ mod tests {
         .await
     }
 
+    // this test is a duplicate of its counterpart but taking federation into account
+    // The heavy lifting of cross-signing the certificates is being done by the test utils.
+    #[async_std::test]
+    #[wasm_bindgen_test]
+    async fn should_read_revoked_device_cross_signed() {
```
Comment (istankovic) [inline crypto/src/e2e_identity/identity.rs:266]: This test is exactly the same as `should_read_revoked_device`, the only thing different looks to be the name. It feels like something is missing.

File: crypto/src/e2e_identity/identity.rs [hunk 9]
```diff
@@ -421,6 +529,119 @@ mod tests {
...
+                    let all_identities = alice_android_central
+                        .mls_central
+                        .get_user_identities(&id, &[alice_user_id.clone(), bob_user_id.clone()])
+                        .await
+                        .unwrap();
+                    assert_eq!(all_identities.len(), 2);
+                    let alice_identities = alice_identities.get(&alice_user_id).unwrap();
+                    assert_eq!(alice_identities.len(), 2);
+                    let bob_identities = bob_identities.get(&bob_user_id).unwrap();
+                    assert_eq!(bob_identities.len(), 1);
```
Comment (istankovic) [inline crypto/src/e2e_identity/identity.rs:626]: This looks weird. And I realize this is just copied from the original test, but I think we should nevertheless address it.

Here we just check that the length of `all_identities` is 2.
But then, we repeat the same checks for Bob and Alice that we had above. So I'm not really sure what this is supposed to test. Maybe the original author intended to write something like the following?

```rust
let all_identities = ...
assert_eq!(all_identities.get(&alice_user_id).unwrap().len(), 2);
assert_eq!(all_identities.get(&bob_user_id).unwrap().len(), 1);
```
That is, to use `all_identities` for the checks.

File: crypto/src/e2e_identity/identity.rs [hunk 10]
```diff
@@ -520,4 +741,83 @@ mod tests {
...
+                        .try_talk_to(&id, &mut alices_ios_central)
+                        .await
+                        .unwrap();
+
+                    // same server communication
+                    bob_android_central
+                        .try_talk_to(&id, &mut alices_ios_central)
+                        .await
+                        .unwrap();
+                })
```
Comment (istankovic) [inline crypto/src/e2e_identity/identity.rs:818]: This is actually an important test, happy to see it! :-)

**Review by istankovic — CHANGES_REQUESTED** [review]
There's still some work to do.

File: crypto/src/e2e_identity/identity.rs [hunk 11]
```diff
@@ -156,20 +156,18 @@ impl MlsConversation {
...
                 let uid = String::try_from(uid);
-                let identities = group.into_iter().map(|(_, id)| id).collect::<CryptoResult<Vec<_>>>();
                 // could be simplified if `Result::zip` was available
-                uid.and_then(|uid| identities.map(|ids| (uid, ids)))
+                uid.and_then(|uid| identity.map(|id| (uid, id)))
             })
-            .collect::<CryptoResult<HashMap<_, _>>>()
+            .process_results(|iter| iter.into_group_map())
     }
 }
```
Comment (istankovic) [inline crypto/src/e2e_identity/identity.rs:172]: Could you please factor this out into a separate commit, so that it is not mixed in with a bunch of unrelated changes? It would be good to mention in the commit message what was wrong with previous code.

**Review by istankovic — APPROVED** [review]
Looks good. :+1:

File: crypto/src/e2e_identity/identity.rs [hunk 12]
```diff
@@ -186,6 +190,54 @@ mod tests {
...
+            .get_user_identities(id, &["aaaaaaaaaaaaa".to_string()])
+            .await
+            .unwrap();
+        assert!(not_found.is_empty());
+
+        // Invalid usage
+        let invalid = central.get_user_identities(id, &[]).await;
+        assert!(matches!(invalid.unwrap_err(), CryptoError::ConsumerError));
+    }
+    async fn check_identities_device_status<const N: usize>(
```
Comment (istankovic) [inline crypto/src/e2e_identity/identity.rs:213]: Missing blank line above this one.
